File: aiworkbench_ui-bugfix_735_ui/aiworkbench_ui-bugfix_735_ui/src/constants/estimateOptions.ts
export const ESTIMATE_LINE_KEYS = [
  'infra',
  'build',
  'validation',
  'support',
  'llm_token',
  'change_management',
] as const;

export type EstimateLineKey = (typeof ESTIMATE_LINE_KEYS)[number];

export const ESTIMATE_LINE_LABELS: Record<EstimateLineKey, string> = {
  infra: 'Infra Cost',
  build: 'Build Cost',
  validation: 'Validation/Testing Cost',
  support: 'Support Cost',
  llm_token: 'LLM Token Cost',
  change_management: 'Change Management Cost',
};

export const ESTIMATE_LINE_TYPES: Record<EstimateLineKey, string[]> = {
  infra: ['Cloud', 'On-Premise'],
  build: ['Inhouse', 'Vendor'],
  validation: ['Internal', 'External'],
  support: ['Inhouse', 'Vendor'],
  llm_token: ['Claude', 'Azure', 'Others'],
  change_management: ['High', 'Medium', 'Low'],
};

export const CURRENCY_OPTIONS = ['USD', 'EUR', 'GBP', 'INR', 'AED', 'SGD', 'JPY', 'AUD', 'CAD'];

/** Recommended vendor evaluation questions when Build Cost type is Vendor. */
export const VENDOR_ASSESSMENT_QUESTIONS = [
  {
    id: 'vendor_legal_entity',
    question: 'Is the vendor a legally registered entity with a named contracting party and jurisdiction?',
  },
  {
    id: 'vendor_security_certs',
    question: 'Does the vendor hold current security certifications (e.g. ISO 27001, SOC 2) relevant to this use case?',
  },
  {
    id: 'vendor_data_residency',
    question: 'Can the vendor confirm data residency, processing locations, and subprocessors for this solution?',
  },
  {
    id: 'vendor_no_train_on_data',
    question: 'Will customer data and prompts be excluded from model training unless explicitly contracted?',
  },
  {
    id: 'vendor_privacy_dpa',
    question: 'Is a Data Processing Agreement (or equivalent) available covering confidentiality, retention, and deletion?',
  },
  {
    id: 'vendor_access_control',
    question: 'Does the vendor support least-privilege access, SSO/MFA, and audit logs for operator activity?',
  },
  {
    id: 'vendor_model_transparency',
    question: 'Can the vendor describe the model/provider stack, update cadence, and known limitations?',
  },
  {
    id: 'vendor_bias_testing',
    question: 'Has the vendor documented bias, fairness, or safety testing for the intended use?',
  },
  {
    id: 'vendor_sla_support',
    question: 'Are SLA, incident response, and support hours defined for production use?',
  },
  {
    id: 'vendor_ip_exit',
    question: 'Are IP ownership of outputs and an exit plan (data return/deletion) contractually defined?',
  },
] as const;

export interface EstimateLine {
  type: string | null;
  actual_cost: number | null;
  year1: number | null;
  year2: number | null;
  year3: number | null;
}

export interface EstimateData {
  currency: string;
  lines: Record<EstimateLineKey, EstimateLine>;
  totals: { year1: number; year2: number; year3: number };
  vendor_checklist: Record<string, 'yes' | 'no' | 'na'>;
}

export function emptyEstimateLine(): EstimateLine {
  return { type: null, actual_cost: null, year1: null, year2: null, year3: null };
}

export function emptyEstimateData(currency = 'USD'): EstimateData {
  const lines = {} as Record<EstimateLineKey, EstimateLine>;
  for (const key of ESTIMATE_LINE_KEYS) {
    lines[key] = emptyEstimateLine();
  }
  return { currency, lines, totals: { year1: 0, year2: 0, year3: 0 }, vendor_checklist: {} };
}

export function isVendorBuild(data: EstimateData | null | undefined): boolean {
  return (data?.lines?.build?.type || '').toLowerCase() === 'vendor';
}

export function computeEstimateTotals(lines: Record<EstimateLineKey, EstimateLine>) {
  const totals = { year1: 0, year2: 0, year3: 0 };
  for (const key of ESTIMATE_LINE_KEYS) {
    const line = lines[key] || emptyEstimateLine();
    for (const year of ['year1', 'year2', 'year3'] as const) {
      const val = line[year];
      totals[year] += typeof val === 'number' && !Number.isNaN(val) ? val : 0;
    }
  }
  return {
    year1: Math.round(totals.year1 * 100) / 100,
    year2: Math.round(totals.year2 * 100) / 100,
    year3: Math.round(totals.year3 * 100) / 100,
  };
}

export const MAX_ESTIMATE_AMOUNT = 999_999_999_999.99;

export type MoneyField = 'actual_cost' | 'year1' | 'year2' | 'year3';

export const MONEY_FIELD_LABELS: Record<MoneyField, string> = {
  actual_cost: 'Actual cost',
  year1: 'Year 1',
  year2: 'Year 2',
  year3: 'Year 3',
};

export function validateCurrency(currency: string, required = false): string | null {
  const value = (currency || '').trim().toUpperCase();
  if (!value) return required ? 'Select a currency.' : null;
  if (!CURRENCY_OPTIONS.includes(value)) return 'Choose a valid currency.';
  return null;
}

export function sanitizeMoneyTyping(raw: string): string {
  const cleaned = raw.replace(/[,\s]/g, '').replace(/[^\d.]/g, '');
  const firstDot = cleaned.indexOf('.');
  if (firstDot === -1) return cleaned.slice(0, 12);
  const whole = cleaned.slice(0, firstDot).slice(0, 12);
  const fraction = cleaned
    .slice(firstDot + 1)
    .replace(/\./g, '')
    .slice(0, 2);
  return `${whole}.${fraction}`;
}

export function parseMoneyInput(raw: string): { value: number | null; error: string | null } {
  const trimmed = raw.trim().replace(/,/g, '');
  if (trimmed === '' || trimmed === '.') return { value: null, error: null };
  if (!/^\d+(\.\d{1,2})?$/.test(trimmed)) {
    if (/^\d+\.$/.test(trimmed)) return { value: null, error: null };
    return { value: null, error: 'Enter a valid amount.' };
  }
  const value = Number(trimmed);
  return { value, error: validateMoneyValue(value, false) };
}

export function validateMoneyValue(value: number | null, required: boolean): string | null {
  if (value == null) return required ? 'This amount is required.' : null;
  if (!Number.isFinite(value)) return 'Enter a valid number.';
  if (value < 0) return 'Amount cannot be negative.';
  if (value > MAX_ESTIMATE_AMOUNT) return 'Amount is too large.';
  const cents = Math.round(value * 100);
  if (Math.abs(value * 100 - cents) > 1e-6) return 'Use at most 2 decimal places.';
  return null;
}

export function lineFieldErrors(
  key: EstimateLineKey,
  line: EstimateLine,
  required: boolean
): Partial<Record<'type' | MoneyField, string>> {
  const errors: Partial<Record<'type' | MoneyField, string>> = {};
  const allowed = ESTIMATE_LINE_TYPES[key];
  if (required && !line.type) errors.type = 'Select a type.';
  else if (line.type && !allowed.includes(line.type)) errors.type = 'Choose a valid type.';
  for (const field of ['actual_cost', 'year1', 'year2', 'year3'] as const) {
    const err = validateMoneyValue(line[field], required);
    if (err) errors[field] = err;
  }
  return errors;
}

export function estimateLineIsComplete(key: EstimateLineKey, line: EstimateLine): boolean {
  return Object.keys(lineFieldErrors(key, line, true)).length === 0;
}

export function estimateLineHasInvalidValues(key: EstimateLineKey, line: EstimateLine): boolean {
  return Object.keys(lineFieldErrors(key, line, false)).length > 0;
}

export function lineYearTotal(line: EstimateLine): number {
  return [line.year1, line.year2, line.year3].reduce(
    (sum, value) => sum + (typeof value === 'number' && Number.isFinite(value) ? value : 0),
    0
  );
}

export type LineNavTone = 'ok' | 'error' | 'warn' | 'idle';

export interface LineNavState {
  tone: LineNavTone;
  hint: string;
  issueCount: number;
}

export function lineNavState(
  key: EstimateLineKey,
  data: EstimateData,
  showWarnings: boolean
): LineNavState {
  const line = data.lines[key] || emptyEstimateLine();
  const invalid = lineFieldErrors(key, line, false);
  const missing = lineFieldErrors(key, line, true);
  const invalidMessages = Object.values(invalid);
  const missingCount = Object.keys(missing).length;

  let extra = 0;
  let extraHint = '';
  if (key === 'build' && isVendorBuild(data)) {
    extra = VENDOR_ASSESSMENT_QUESTIONS.filter((q) => {
      const v = data.vendor_checklist[q.id];
      return v !== 'yes' && v !== 'no' && v !== 'na';
    }).length;
    if (extra) extraHint = `${extra} vendor question${extra === 1 ? '' : 's'} unanswered`;
  }

  if (invalidMessages.length) {
    return {
      tone: 'error',
      hint: extraHint ? `${invalidMessages[0]} · ${extraHint}` : invalidMessages[0],
      issueCount: invalidMessages.length + extra,
    };
  }

  if (missingCount === 0 && extra === 0) {
    return {
      tone: 'ok',
      hint: line.type || 'Complete',
      issueCount: 0,
    };
  }

  const parts: string[] = [];
  if (missing.type) parts.push('Select a type');
  const amountMissing = (['actual_cost', 'year1', 'year2', 'year3'] as const).filter((field) => missing[field]);
  if (amountMissing.length === 4) parts.push('Enter all amounts');
  else if (amountMissing.length === 1) parts.push(`${MONEY_FIELD_LABELS[amountMissing[0]]} required`);
  else if (amountMissing.length > 1) parts.push(`${amountMissing.length} amounts required`);
  if (extraHint) parts.push(extraHint);

  const hint = parts.join(' · ') || 'Incomplete';
  if (showWarnings) {
    return { tone: 'warn', hint, issueCount: missingCount + extra };
  }
  return {
    tone: 'idle',
    hint: line.type || 'Type not selected',
    issueCount: missingCount + extra,
  };
}

export function estimateProgress(data: EstimateData): { done: number; total: number } {
  const done = ESTIMATE_LINE_KEYS.filter((key) => estimateLineIsComplete(key, data.lines[key])).length;
  return { done, total: ESTIMATE_LINE_KEYS.length };
}

export function vendorChecklistProgress(data: EstimateData): { done: number; total: number } {
  const total = VENDOR_ASSESSMENT_QUESTIONS.length;
  const done = VENDOR_ASSESSMENT_QUESTIONS.filter((q) => {
    const v = data.vendor_checklist[q.id];
    return v === 'yes' || v === 'no' || v === 'na';
  }).length;
  return { done, total };
}

export function estimateFilledValueErrors(data: EstimateData): string | null {
  const currencyErr = validateCurrency(data.currency, false);
  if (currencyErr) return currencyErr;
  for (const key of ESTIMATE_LINE_KEYS) {
    const line = data.lines[key] || emptyEstimateLine();
    const errors = lineFieldErrors(key, line, false);
    const first = Object.values(errors)[0];
    if (first) return `${ESTIMATE_LINE_LABELS[key]}: ${first}`;
  }
  if (isVendorBuild(data)) {
    const invalid = VENDOR_ASSESSMENT_QUESTIONS.some((q) => {
      const v = data.vendor_checklist[q.id];
      return v != null && v !== 'yes' && v !== 'no' && v !== 'na';
    });
    if (invalid) return 'Vendor assessment answers must be Yes, No, or N/A.';
  }
  return null;
}

export function estimateCompletenessError(data: EstimateData): string | null {
  const filledErr = estimateFilledValueErrors(data);
  if (filledErr) return filledErr;
  const currencyErr = validateCurrency(data.currency, true);
  if (currencyErr) return currencyErr;
  for (const key of ESTIMATE_LINE_KEYS) {
    const line = data.lines[key] || emptyEstimateLine();
    const errors = lineFieldErrors(key, line, true);
    const first = Object.values(errors)[0];
    if (first) return `${ESTIMATE_LINE_LABELS[key]}: ${first}`;
  }
  if (isVendorBuild(data)) {
    const unanswered = VENDOR_ASSESSMENT_QUESTIONS.some((q) => {
      const v = data.vendor_checklist[q.id];
      return v !== 'yes' && v !== 'no' && v !== 'na';
    });
    if (unanswered) return 'Answer all vendor assessment questions (Yes / No / N/A).';
  }
  return null;
}

export function estimateMissingHints(data: EstimateData): string[] {
  const hints: string[] = [];
  for (const key of ESTIMATE_LINE_KEYS) {
    const line = data.lines[key] || emptyEstimateLine();
    const errors = lineFieldErrors(key, line, true);
    if (errors.type) hints.push(`${ESTIMATE_LINE_LABELS[key]}: type`);
    for (const field of ['actual_cost', 'year1', 'year2', 'year3'] as const) {
      if (errors[field]) hints.push(`${ESTIMATE_LINE_LABELS[key]}: ${MONEY_FIELD_LABELS[field].toLowerCase()}`);
    }
  }
  if (isVendorBuild(data)) {
    const left = VENDOR_ASSESSMENT_QUESTIONS.filter((q) => {
      const v = data.vendor_checklist[q.id];
      return v !== 'yes' && v !== 'no' && v !== 'na';
    }).length;
    if (left) hints.push(`${left} vendor question${left === 1 ? '' : 's'}`);
  }
  return hints;
}

export function firstIncompleteLine(data: EstimateData): EstimateLineKey | null {
  return ESTIMATE_LINE_KEYS.find((key) => !estimateLineIsComplete(key, data.lines[key])) || null;
}

export function normalizeEstimateData(raw: any, defaultCurrency = 'USD'): EstimateData {
  const base = emptyEstimateData(
    CURRENCY_OPTIONS.includes((raw?.currency || defaultCurrency || 'USD').toUpperCase())
      ? (raw?.currency || defaultCurrency).toUpperCase()
      : defaultCurrency
  );
  if (!raw || typeof raw !== 'object') return base;
  const linesIn = raw.lines && typeof raw.lines === 'object' ? raw.lines : {};
  for (const key of ESTIMATE_LINE_KEYS) {
    const src = linesIn[key] && typeof linesIn[key] === 'object' ? linesIn[key] : {};
    const allowed = ESTIMATE_LINE_TYPES[key];
    const lineType = allowed.includes(src.type) ? src.type : null;
    const parseYear = (v: unknown) => {
      if (v === null || v === undefined || v === '') return null;
      const n = Number(v);
      return Number.isFinite(n) ? n : null;
    };
    base.lines[key] = {
      type: lineType,
      actual_cost: parseYear(src.actual_cost),
      year1: parseYear(src.year1),
      year2: parseYear(src.year2),
      year3: parseYear(src.year3),
    };
  }
  base.totals = computeEstimateTotals(base.lines);
  const checklist = raw.vendor_checklist && typeof raw.vendor_checklist === 'object' ? raw.vendor_checklist : {};
  const normalizedChecklist: Record<string, 'yes' | 'no' | 'na'> = {};
  for (const q of VENDOR_ASSESSMENT_QUESTIONS) {
    const v = checklist[q.id];
    if (v === true || v === 'yes') normalizedChecklist[q.id] = 'yes';
    else if (v === false || v === 'no') normalizedChecklist[q.id] = 'no';
    else if (v === 'na') normalizedChecklist[q.id] = 'na';
  }
  base.vendor_checklist = normalizedChecklist;
  return base;
}

File: aiworkbench_api/app/core/estimate_options.py
"""
Estimate-phase cost option lists and helpers.
"""

from decimal import Decimal, InvalidOperation, ROUND_HALF_UP

INFRA_COST_TYPES = ["Cloud", "On-Premise"]
BUILD_COST_TYPES = ["Inhouse", "Vendor"]
VALIDATION_COST_TYPES = ["Internal", "External"]
SUPPORT_COST_TYPES = ["Inhouse", "Vendor"]
LLM_TOKEN_COST_TYPES = ["Claude", "Azure", "Others"]
CHANGE_MGMT_COST_TYPES = ["High", "Medium", "Low"]

CURRENCY_OPTIONS = ["USD", "EUR", "GBP", "INR", "AED", "SGD", "JPY", "AUD", "CAD"]

ESTIMATE_LINE_KEYS = [
    "infra",
    "build",
    "validation",
    "support",
    "llm_token",
    "change_management",
]

ESTIMATE_LINE_LABELS = {
    "infra": "Infra Cost",
    "build": "Build Cost",
    "validation": "Validation/Testing Cost",
    "support": "Support Cost",
    "llm_token": "LLM Token Cost",
    "change_management": "Change Management Cost",
}

ESTIMATE_LINE_TYPES = {
    "infra": INFRA_COST_TYPES,
    "build": BUILD_COST_TYPES,
    "validation": VALIDATION_COST_TYPES,
    "support": SUPPORT_COST_TYPES,
    "llm_token": LLM_TOKEN_COST_TYPES,
    "change_management": CHANGE_MGMT_COST_TYPES,
}

VENDOR_ANSWER_VALUES = ("yes", "no", "na", True, False)

MAX_ESTIMATE_AMOUNT = Decimal("999999999999.99")
TWO_PLACES = Decimal("0.01")


def empty_estimate_line() -> dict:
    return {"type": None, "actual_cost": None, "year1": None, "year2": None, "year3": None}


VENDOR_ASSESSMENT_QUESTIONS = [
    {
        "id": "vendor_legal_entity",
        "question": "Is the vendor a legally registered entity with a named contracting party and jurisdiction?",
    },
    {
        "id": "vendor_security_certs",
        "question": "Does the vendor hold current security certifications (e.g. ISO 27001, SOC 2) relevant to this use case?",
    },
    {
        "id": "vendor_data_residency",
        "question": "Can the vendor confirm data residency, processing locations, and subprocessors for this solution?",
    },
    {
        "id": "vendor_no_train_on_data",
        "question": "Will customer data and prompts be excluded from model training unless explicitly contracted?",
    },
    {
        "id": "vendor_privacy_dpa",
        "question": "Is a Data Processing Agreement (or equivalent) available covering confidentiality, retention, and deletion?",
    },
    {
        "id": "vendor_access_control",
        "question": "Does the vendor support least-privilege access, SSO/MFA, and audit logs for operator activity?",
    },
    {
        "id": "vendor_model_transparency",
        "question": "Can the vendor describe the model/provider stack, update cadence, and known limitations?",
    },
    {
        "id": "vendor_bias_testing",
        "question": "Has the vendor documented bias, fairness, or safety testing for the intended use?",
    },
    {
        "id": "vendor_sla_support",
        "question": "Are SLA, incident response, and support hours defined for production use?",
    },
    {
        "id": "vendor_ip_exit",
        "question": "Are IP ownership of outputs and an exit plan (data return/deletion) contractually defined?",
    },
]


def is_vendor_build(data: dict | None) -> bool:
    if not isinstance(data, dict):
        return False
    lines = data.get("lines") if isinstance(data.get("lines"), dict) else {}
    build = lines.get("build") if isinstance(lines.get("build"), dict) else {}
    return str(build.get("type") or "").strip().lower() == "vendor"


def _to_decimal(value: object) -> Decimal | None:
    if value in (None, ""):
        return None
    if isinstance(value, bool):
        raise InvalidOperation("boolean is not an amount")
    text = str(value).strip().replace(",", "")
    if text == "":
        return None
    number = Decimal(text)
    if not number.is_finite():
        raise InvalidOperation("non-finite")
    return number


def _amount_error(value: object, label: str, required: bool = False) -> str | None:
    try:
        number = _to_decimal(value)
    except (InvalidOperation, TypeError, ValueError):
        return f"{label} must be a valid number."
    if number is None:
        return f"{label} is required." if required else None
    if number < 0:
        return f"{label} cannot be negative."
    if number > MAX_ESTIMATE_AMOUNT:
        return f"{label} is too large."
    quantized = number.quantize(TWO_PLACES, rounding=ROUND_HALF_UP)
    if (number - quantized).copy_abs() > Decimal("0.001"):
        return f"{label} can have at most 2 decimal places."
    return None


def _normalize_vendor_answer(value: object) -> str | None:
    if value is True or value == "yes":
        return "yes"
    if value is False or value == "no":
        return "no"
    if value == "na":
        return "na"
    return None


def validate_estimate_data(data: dict | None, *, require_complete: bool = False) -> str | None:
    """Return first validation error, or None if valid."""
    if not isinstance(data, dict):
        return "Estimate data is required." if require_complete else None

    currency = str(data.get("currency") or "").strip().upper()
    if currency and currency not in CURRENCY_OPTIONS:
        return "Choose a valid currency."
    if require_complete and not currency:
        return "Currency is required."

    lines = data.get("lines") if isinstance(data.get("lines"), dict) else {}
    for key in ESTIMATE_LINE_KEYS:
        line = lines.get(key) if isinstance(lines.get(key), dict) else {}
        label = ESTIMATE_LINE_LABELS.get(key, key)
        allowed = ESTIMATE_LINE_TYPES[key]
        line_type = line.get("type")
        if require_complete and not line_type:
            return f"{label}: type is required."
        if line_type and line_type not in allowed:
            return f"{label}: choose a valid type."
        for field, field_label in (
            ("year1", "Year 1"),
            ("year2", "Year 2"),
            ("year3", "Year 3"),
        ):
            err = _amount_error(
                line.get(field),
                f"{label} {field_label}",
                required=require_complete,
            )
            if err:
                return err
        leftover_actual = line.get("actual_cost")
        if leftover_actual not in (None, ""):
            leftover_err = _amount_error(leftover_actual, f"{label} actual cost", required=False)
            if leftover_err:
                return leftover_err

    checklist = data.get("vendor_checklist") if isinstance(data.get("vendor_checklist"), dict) else {}
    for item in VENDOR_ASSESSMENT_QUESTIONS:
        raw = checklist.get(item["id"], None)
        if raw in (None, ""):
            continue
        if _normalize_vendor_answer(raw) is None:
            return "Vendor assessment answers must be Yes, No, or N/A."
    return None


def empty_estimate_payload(currency: str = "USD") -> dict:
    return {
        "currency": currency,
        "lines": {key: empty_estimate_line() for key in ESTIMATE_LINE_KEYS},
        "totals": {"year1": 0, "year2": 0, "year3": 0},
        "vendor_checklist": {},
    }


def compute_estimate_totals(lines: dict) -> dict:
    totals = {"year1": Decimal("0"), "year2": Decimal("0"), "year3": Decimal("0")}
    for key in ESTIMATE_LINE_KEYS:
        line = lines.get(key) or {}
        for year in ("year1", "year2", "year3"):
            try:
                amount = _to_decimal(line.get(year))
            except (InvalidOperation, TypeError, ValueError):
                amount = None
            if amount is not None and amount >= 0:  
                totals[year] += amount.quantize(TWO_PLACES, rounding=ROUND_HALF_UP)
    return {k: float(v.quantize(TWO_PLACES, rounding=ROUND_HALF_UP)) for k, v in totals.items()}


def _parse_amount(value: object) -> float | None:
    try:
        number = _to_decimal(value)
    except (InvalidOperation, TypeError, ValueError):
        return None
    if number is None:
        return None
    quantized = number.quantize(TWO_PLACES, rounding=ROUND_HALF_UP)
    if (number - quantized).copy_abs() > Decimal("0.001"):
        return float(number)
    return float(quantized)


def normalize_estimate_data(raw: dict | None, default_currency: str = "USD") -> dict:
    base = empty_estimate_payload(default_currency)
    if not isinstance(raw, dict):
        return base
    currency = (raw.get("currency") or default_currency or "USD").strip().upper()
    if currency not in CURRENCY_OPTIONS:
        currency = default_currency if default_currency in CURRENCY_OPTIONS else "USD"
    lines_in = raw.get("lines") if isinstance(raw.get("lines"), dict) else {}
    lines = {}
    for key in ESTIMATE_LINE_KEYS:
        src = lines_in.get(key) if isinstance(lines_in.get(key), dict) else {}
        line_type = src.get("type")
        allowed = ESTIMATE_LINE_TYPES[key]
        if line_type not in allowed:
            line_type = None
        lines[key] = {
            "type": line_type,
            "actual_cost": _parse_amount(src.get("actual_cost")),
            "year1": _parse_amount(src.get("year1")),
            "year2": _parse_amount(src.get("year2")),
            "year3": _parse_amount(src.get("year3")),
        }
    totals = compute_estimate_totals(lines)
    vendor_in = raw.get("vendor_checklist") if isinstance(raw.get("vendor_checklist"), dict) else {}
    vendor_checklist = {}
    for item in VENDOR_ASSESSMENT_QUESTIONS:
        answer = _normalize_vendor_answer(vendor_in.get(item["id"]))
        if answer is not None:
            vendor_checklist[item["id"]] = answer
    return {
        "currency": currency,
        "lines": lines,
        "totals": totals,
        "vendor_checklist": vendor_checklist,
    }
